vhdl/quartus/qsys_templates.py

```diff
@@ -0,0 +1,101 @@
+set_connection_parameter_value axi_master_name/component_name_0.avalon_slave baseAddress {component_base_address}
+set_connection_parameter_value axi_master_name/component_name_0.avalon_slave defaultConnection {0}
+'''
+custom_component_initial_audio_in_connection_template='''
+add_connection audio_in component_name_0.avalon_streaming_sink
+'''
+custom_component_audio_in_connection_template='''
+add_connection last_component_name.avalon_streaming_source component_name_0.avalon_streaming_sink
+'''
+custom_component_final_audio_out_connection_template='''
+add_connection component_name_0.avalon_streaming_source audio_out
+'''
+
+custom_component_instantiation_template = "add_instance component_name_0 component_name 1.0\n"
+
+created_by_function_header_template = """
+# # # # # # # # # # # # # # # # # # # # # #
+# Created by function
+# # # # # # # # # # # # # # # # # # # # # #
+"""
+
+quartus_project_template = """
+set project audio_effect
+set target_system target_name
+load_package flow
+
+project_new $project -revision $target_system -overwrite
+
+source ${target_system}_proj.tcl
+
+#Add files here
+files_list
+
+# compile the project
+execute_flow -compile
+project_close
+"""
+
+class qsys_templates:
+    custom_component_header =(
+        "# # # # # # # # # # # # # # # # # # # # # #\n"
+        "# Custom Components \n"
+        "# # # # # # # # # # # # # # # # # # # # # #\n\n"
+        )
+    def __init__(self, num_custom_components, baseAddress):
+        self.custom_components_added = 0
+        self.num_custom_components = num_custom_components
+        self.de10_components_base_address = 20 # Format is 0x0020
+        self.last_component_added = ""
+    def add_custom_component_instantiaion(self, component_name):
+        return custom_component_instantiation_template.replace("component_name", component_name)
+
+    def add_custom_component_connections(self, component_name, target):
+        # Increment address by 0x10
+        component_base_address = "0x00" + str(self.de10_components_base_address + self.custom_components_added * 10) 
+        
+        built_string = custom_component_base_connections_template \
+        .replace("component_name", component_name) \
+        .replace("component_base_address", component_base_address) \
+        .replace("axi_master_name", target.axi_master_name) \
+        .replace("clock_name", target.clock_name)
+        
+        if(self.custom_components_added == 0):
+            built_string += qsys_templates.custom_component_header
+            built_string += custom_component_initial_audio_in_connection_template.replace("component_name", component_name).replace("audio_in", target.audio_in)
+        if((self.custom_components_added + 1) == self.num_custom_components):
+            built_string += custom_component_final_audio_out_connection_template.replace("component_name", component_name).replace("audio_out", target.audio_out)
+            self.custom_components_added = 0        
```

Comment:
Does this mean that we can add a chain of custom components, e.g. codec -> echo -> gain -> compression -> codec? If so, how does that work?

Reply:
That is what it means. I haven't actually done any upstream work to actually do such a thing, but all that needs to be done for this is adding the custom_component names in order to `input_struct.custom_components`. Currently I am just making an array out of the model name from the JSON.

Reply:
Sweet. That'll be really cool. I suspect adding that into the simulink models themselves won't be worthwhile as we're targeting a single Simulink model scenario for end-users, but having that ability will be cool. I did something similar for combining multiple models into a unified GUI config file. For connecting multiple models, I think we could just take multiple dataplane json files as command-line input to this program. Then people have the option to run the tool manually to achieve the multiple model scenario. Or we could supply some sort of master config file that specifies tool/workflow options, such as combining multiple models, uploading the build files to AWS, etc.